zerver/tests/test_service_bot_system.py

```diff
@@ -118,6 +121,85 @@ def test_service_events_for_unsubscribed_stream_mentions(self):
+                                                 realm=get_realm("zulip"),
+                                                 full_name="EmbeddedBot2",
+                                                 short_name="embedded-bot-2",
+                                                 bot_type=UserProfile.EMBEDDED_BOT,
+                                                 bot_owner=self.user_profile)
+
+    def test_basic_storage_and_retrieval(self):
+        # type: () -> None
+        state_handler = StateHandler(self.bot_profile)
+        state_handler['some key'] = 'some value'
+        state_handler['some other key'] = 'some other value'
+        self.assertEqual(state_handler['some key'], 'some value')
+        self.assertEqual(state_handler['some other key'], 'some other value')
+        self.assertFalse('nonexistent key' in state_handler)
+        self.assertRaises(BotUserStateData.DoesNotExist, lambda: state_handler['nonexistent key'])
+
+        second_state_handler = StateHandler(self.second_bot_profile)
+        self.assertRaises(BotUserStateData.DoesNotExist, lambda: second_state_handler['some key'])
+        second_state_handler['some key'] = 'yet another value'
+        self.assertEqual(state_handler['some key'], 'some value')
+        self.assertEqual(second_state_handler['some key'], 'yet another value')
+
+    def test_marshalling(self):
+        # type: () -> None
+        state_handler = StateHandler(self.bot_profile)
+        serializable_obj = {'foo': 'bar', 'baz': [42, 'cux']}
+        with state_handler(marshal=lambda obj: json.dumps(obj),
+                           demarshal=lambda obj: json.loads(obj)) as state:
+            state['some key'] = serializable_obj  # type: ignore # Ignore for testing.
+            self.assertEqual(state['some key'], serializable_obj)
+
+        def assignment_after_marshalling():
+            # type: () -> None
+            state_handler['some key'] = serializable_obj  # type: ignore # Ignore for testing.
+        self.assertRaises(TypeError, assignment_after_marshalling)
+
+    def test_invalid_calls(self):
+        # type: () -> None
+        state_handler = StateHandler(self.bot_profile)
+        serializable_obj = {'foo': 'bar', 'baz': [42, 'cux']}
+
+        def invalid_value_assignment():
+            # type: () -> None
+            state_handler['some key'] = serializable_obj  # type: ignore # Ignore for testing.
+        self.assertRaises(TypeError, invalid_value_assignment)
+
+        def invalid_key_assignment():
+            # type: () -> None
+            state_handler[serializable_obj] = 'some value'  # type: ignore # Ignore for testing.
+        self.assertRaises(TypeError, invalid_key_assignment)
+
+    def test_storage_limit(self):
+        # type: () -> None
+        state_handler = StateHandler(self.bot_profile)
+        key = 'capacity-filling entry'
+        state_handler[key] = 'x' * (state_handler.state_size_limit - len(key))
```

Comment:
For this test, it'd be a lot better to change `state_handler.state_size_limit` to be like 10 bytes for the test; accessing a 1M entry thing  might make this test slow.

Reply:
FWIW, the current size limit is 10,000,000 bytes.

Reply:
Just to be clear, I'm just suggesting that _for this test_ we want to monkey-patch things to make the size limit small, to keep the test fast.  We do this for limits all over the place: change the limit to like 10 bytes to exercise the code path, rather than using a 10MB object.

Reply:
Yeah, I updated the PR to to use 100 bytes (so we can still use some reasonably verbose key and value strings).